com.microsoft.mrtk.input/Visualizers/ControllerVisualizer/ControllerVisualizer.cs

```diff
@@ -148,19 +155,41 @@ private async void InstantiateControllerVisuals(bool usePlatformVisuals, bool us
+                // Only try to load models from the platform sdk if the controllerGameObject is now null or if
+                bool existsInCache = inputDevicePlatformVisualsCache.TryGetValue(inputDevice, out var results);
+                var (loadableFromPlatform, cachedControllerGameObject) = existsInCache ? results : (false, null);
+
+                if (!existsInCache || (loadableFromPlatform && cachedControllerGameObject == null))
                 {
-                    // Platform models are "rotated" 180 degrees because their forward vector points towards the user.
-                    // We need to rotate these models in order to have them pointing in the correct direction on device.
-                    if (platformLoadedGameObjectRoot == null)
+                    platformLoadedGameObject = await ControllerModelLoader.TryGenerateControllerModelFromPlatformSDK(handNode.ToHandedness());
+                    if (platformLoadedGameObject != null)
                     {
-                        platformLoadedGameObjectRoot = new GameObject("Platform Model Root");
+                        // Platform models are "rotated" 180 degrees because their forward vector points towards the user.
+                        // We need to rotate these models in order to have them pointing in the correct direction on device.
+                        if (platformLoadedGameObjectRoot == null)
+                        {
+                            platformLoadedGameObjectRoot = new GameObject("Platform Model Root");
+                        }
+                        platformLoadedGameObject.transform.parent = platformLoadedGameObjectRoot.transform;
+                        platformLoadedGameObject.transform.SetPositionAndRotation(platformLoadedGameObjectRoot.transform.position, platformLoadedGameObjectRoot.transform.rotation * controllerModelRotatedOffset);
+
+                        controllerGameObject = platformLoadedGameObjectRoot;
                     }
-                    platformLoadedGameObject.transform.parent = platformLoadedGameObjectRoot.transform;
-                    platformLoadedGameObject.transform.SetPositionAndRotation(platformLoadedGameObjectRoot.transform.position, platformLoadedGameObjectRoot.transform.rotation * controllerModelRotatedOffset);
 
-                    controllerGameObject = platformLoadedGameObjectRoot;
+                    // Finally cache whether we were able to load the controller model as well as the model itself
+                    if (!existsInCache)
+                    {
+                        inputDevicePlatformVisualsCache.Add(inputDevice, (platformLoadedGameObject != null, platformLoadedGameObject));
+                    }
+                    else
+                    {
+                        inputDevicePlatformVisualsCache[inputDevice] = (platformLoadedGameObject != null, platformLoadedGameObject);
+                    }
```

Comment:
nit: you could combine this into one line: just the `else` case. The `[inputDevice]` syntax is valid both for adding and replacing.